# **FINELITE** (***AI-powered credit card limit prediction*** )
- **An AI-driven credit card limit prediction system that uses customer demographic, financial, and credit history data to estimate an appropriate credit limit. Developed an end-to-end machine learning pipeline covering data preprocessing, model development, evaluation, explainability, and real-time prediction**

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

**Read the Excel Dataset**

In [2]:
df = pd.read_excel('../Datasets/Credir_Card_Bank.xlsx')
pd.set_option("display.max_columns",None)
df

,Customer_ID,Age,Gender,Employment_Type,Occupation,Monthly_Income,Annual_Income,Credit_Score,Years_With_Bank,Existing_Credit_Cards,Existing_Credit_Limit,Loan_Count,EMI_Per_Month,Debt_To_Income_Ratio,Savings_Balance,Investment_Value,Avg_Monthly_Transactions,Avg_Monthly_Spending,Credit_Utilization,Credit_History_Years,Missed_Payments,Late_Payment_Count,Number_of_Defaults,Residential_Status,PAN_Verified,KYC_Status,Fraud_Flag,Credit_Limit
0,CUST00001,41,Female,Salaried,Sales Executive,140273,1683276,811,4,4,706088,3,49141,0.33,221285,248294,221,69992.0,34,11,0,8,0,Rent,Yes,Complete,No,1373925
1,CUST00002,23,Male,Self-Employed,Business Owner,209521,2514252,837,12,2,527133,1,13338,0.24,348033,1197179,199,17170.0,71,20,3,3,0,Own,Yes,Complete,No,1902449
2,CUST00003,45,Female,Salaried,Designer,47682,572184,679,2,6,281776,1,3618,0.18,603413,1459090,76,16871.0,88,12,5,2,0,Rent,Yes,Complete,No,1165913
3,CUST00004,39,Male,Salaried,Doctor,214966,2579592,660,5,0,656977,0,16114,0.05,1113302,1362180,112,193469.4,66,4,3,5,0,Rent,Yes,Complete,No,1869491
4,CUST00005,58,Male,Self-Employed,Designer,30051,360612,589,18,0,267892,4,36361,0.32,824809,595409,46,6024.0,70,1,4,2,0,Own,Yes,Complete,No,825275
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4995,CUST04996,48,Male,Salaried,Doctor,55078,660936,824,12,1,806871,4,34469,0.10,433388,77558,211,30151.0,91,7,4,4,0,Rent,Yes,Complete,No,1156917
4996,CUST04997,59,Female,Self-Employed,Accountant,107040,1284480,840,0,1,896965,4,40182,0.25,254850,1934214,61,86594.0,28,22,1,6,0,Own,Yes,Complete,No,1458578
4997,CUST04998,26,Female,Salaried,Doctor,168754,2025048,632,19,0,712921,1,52001,0.09,91677,868069,210,42256.0,38,11,3,1,0,Rent,Yes,Complete,No,1342610
4998,CUST04999,42,Male,Salaried,Manager,197110,2365320,734,9,2,91279,1,40678,0.13,1256605,1989152,49,177399.0,15,3,2,2,0,Own,Yes,Complete,No,1868828


**Dataset Overview, Data Types & Missing Value Analysis**

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 28 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Customer_ID               5000 non-null   object 
 1   Age                       5000 non-null   int64  
 2   Gender                    5000 non-null   object 
 3   Employment_Type           5000 non-null   object 
 4   Occupation                5000 non-null   object 
 5   Monthly_Income            5000 non-null   int64  
 6   Annual_Income             5000 non-null   int64  
 7   Credit_Score              5000 non-null   int64  
 8   Years_With_Bank           5000 non-null   int64  
 9   Existing_Credit_Cards     5000 non-null   int64  
 10  Existing_Credit_Limit     5000 non-null   int64  
 11  Loan_Count                5000 non-null   int64  
 12  EMI_Per_Month             5000 non-null   int64  
 13  Debt_To_Income_Ratio      5000 non-null   float64
 14  Savings_

In [4]:
df.isnull().sum()

Customer_ID                 0
Age                         0
Gender                      0
Employment_Type             0
Occupation                  0
Monthly_Income              0
Annual_Income               0
Credit_Score                0
Years_With_Bank             0
Existing_Credit_Cards       0
Existing_Credit_Limit       0
Loan_Count                  0
EMI_Per_Month               0
Debt_To_Income_Ratio        0
Savings_Balance             0
Investment_Value            0
Avg_Monthly_Transactions    0
Avg_Monthly_Spending        0
Credit_Utilization          0
Credit_History_Years        0
Missed_Payments             0
Late_Payment_Count          0
Number_of_Defaults          0
Residential_Status          0
PAN_Verified                0
KYC_Status                  0
Fraud_Flag                  0
Credit_Limit                0
dtype: int64

In [5]:
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


### **Data Preprocessing**
**Correlation Analysis**

In [6]:
df.corr(numeric_only=True)

,Age,Monthly_Income,Annual_Income,Credit_Score,Years_With_Bank,Existing_Credit_Cards,Existing_Credit_Limit,Loan_Count,EMI_Per_Month,Debt_To_Income_Ratio,Savings_Balance,Investment_Value,Avg_Monthly_Transactions,Avg_Monthly_Spending,Credit_Utilization,Credit_History_Years,Missed_Payments,Late_Payment_Count,Number_of_Defaults,Credit_Limit
Age,1.000000,0.000586,0.000586,-0.004417,0.001938,0.033044,0.015468,0.013240,0.018521,-0.001949,-0.005467,0.004425,-0.001134,-0.001287,-0.003089,-0.008412,-0.005983,0.040799,-0.004269,-0.003773
Monthly_Income,0.000586,1.000000,1.000000,-0.014696,-0.022105,-0.009906,0.012751,-0.021474,0.008250,0.001379,-0.005543,-0.022537,0.006890,0.628810,0.009198,0.015196,-0.015982,0.010823,0.021510,0.845154
Annual_Income,0.000586,1.000000,1.000000,-0.014696,-0.022105,-0.009906,0.012751,-0.021474,0.008250,0.001379,-0.005543,-0.022537,0.006890,0.628810,0.009198,0.015196,-0.015982,0.010823,0.021510,0.845154
Credit_Score,-0.004417,-0.014696,-0.014696,1.000000,0.019403,-0.004087,0.013865,0.000986,0.007697,0.003660,-0.017790,0.008831,0.007820,-0.024062,0.010176,0.015656,0.012040,0.005148,-0.356127,0.381760
Years_With_Bank,0.001938,-0.022105,-0.022105,0.019403,1.000000,-0.002723,-0.002130,0.015532,-0.028507,0.003181,-0.008334,0.024473,0.009605,-0.009075,-0.014714,-0.015606,-0.015810,0.013017,0.000524,-0.006032
Existing_Credit_Cards,0.033044,-0.009906,-0.009906,-0.004087,-0.002723,1.000000,0.033060,-0.003890,0.003283,-0.003248,-0.001928,-0.016971,-0.013516,-0.016883,-0.006137,-0.004833,-0.014084,-0.023603,0.001017,-0.013278
Existing_Credit_Limit,0.015468,0.012751,0.012751,0.013865,-0.002130,0.033060,1.000000,0.034080,-0.017020,0.002645,0.000433,0.009725,0.009526,-0.023652,0.036823,-0.019915,-0.020550,0.017083,-0.000667,0.019990
Loan_Count,0.013240,-0.021474,-0.021474,0.000986,0.015532,-0.003890,0.034080,1.000000,-0.020558,0.002716,-0.018963,-0.011324,-0.001872,-0.023220,-0.016886,-0.010709,-0.021134,-0.013663,0.002175,-0.021052
EMI_Per_Month,0.018521,0.008250,0.008250,0.007697,-0.028507,0.003283,-0.017020,-0.020558,1.000000,0.016604,0.023609,-0.010898,0.009866,0.006183,-0.022559,0.009839,-0.003114,-0.014387,-0.013755,-0.156737
Debt_To_Income_Ratio,-0.001949,0.001379,0.001379,0.003660,0.003181,-0.003248,0.002645,0.002716,0.016604,1.000000,-0.014972,-0.015340,-0.013335,-0.016694,-0.013320,0.003050,0.004718,0.007727,-0.012275,-0.108885


**Remove Irrelevant Features & Encode Categorical Variables**

In [7]:
df = df.drop(columns=["Customer_ID","Monthly_Income","PAN_Verified", "KYC_Status"])
df

,Age,Gender,Employment_Type,Occupation,Annual_Income,Credit_Score,Years_With_Bank,Existing_Credit_Cards,Existing_Credit_Limit,Loan_Count,EMI_Per_Month,Debt_To_Income_Ratio,Savings_Balance,Investment_Value,Avg_Monthly_Transactions,Avg_Monthly_Spending,Credit_Utilization,Credit_History_Years,Missed_Payments,Late_Payment_Count,Number_of_Defaults,Residential_Status,Fraud_Flag,Credit_Limit
0,41,Female,Salaried,Sales Executive,1683276,811,4,4,706088,3,49141,0.33,221285,248294,221,69992.0,34,11,0,8,0,Rent,No,1373925
1,23,Male,Self-Employed,Business Owner,2514252,837,12,2,527133,1,13338,0.24,348033,1197179,199,17170.0,71,20,3,3,0,Own,No,1902449
2,45,Female,Salaried,Designer,572184,679,2,6,281776,1,3618,0.18,603413,1459090,76,16871.0,88,12,5,2,0,Rent,No,1165913
3,39,Male,Salaried,Doctor,2579592,660,5,0,656977,0,16114,0.05,1113302,1362180,112,193469.4,66,4,3,5,0,Rent,No,1869491
4,58,Male,Self-Employed,Designer,360612,589,18,0,267892,4,36361,0.32,824809,595409,46,6024.0,70,1,4,2,0,Own,No,825275
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4995,48,Male,Salaried,Doctor,660936,824,12,1,806871,4,34469,0.10,433388,77558,211,30151.0,91,7,4,4,0,Rent,No,1156917
4996,59,Female,Self-Employed,Accountant,1284480,840,0,1,896965,4,40182,0.25,254850,1934214,61,86594.0,28,22,1,6,0,Own,No,1458578
4997,26,Female,Salaried,Doctor,2025048,632,19,0,712921,1,52001,0.09,91677,868069,210,42256.0,38,11,3,1,0,Rent,No,1342610
4998,42,Male,Salaried,Manager,2365320,734,9,2,91279,1,40678,0.13,1256605,1989152,49,177399.0,15,3,2,2,0,Own,No,1868828


In [8]:
df = pd.get_dummies(df,columns=['Gender','Employment_Type','Occupation','Residential_Status','Fraud_Flag'],drop_first=True)
print(df.shape)
df.head()

(5000, 30)


,Age,Annual_Income,Credit_Score,Years_With_Bank,Existing_Credit_Cards,Existing_Credit_Limit,Loan_Count,EMI_Per_Month,Debt_To_Income_Ratio,Savings_Balance,Investment_Value,Avg_Monthly_Transactions,Avg_Monthly_Spending,Credit_Utilization,Credit_History_Years,Missed_Payments,Late_Payment_Count,Number_of_Defaults,Credit_Limit,Gender_Male,Employment_Type_Self-Employed,Occupation_Business Owner,Occupation_Designer,Occupation_Doctor,Occupation_Manager,Occupation_Sales Executive,Occupation_Software Engineer,Occupation_Teacher,Residential_Status_Rent,Fraud_Flag_Yes
0,41,1683276,811,4,4,706088,3,49141,0.33,221285,248294,221,69992.0,34,11,0,8,0,1373925,False,False,False,False,False,False,True,False,False,True,False
1,23,2514252,837,12,2,527133,1,13338,0.24,348033,1197179,199,17170.0,71,20,3,3,0,1902449,True,True,True,False,False,False,False,False,False,False,False
2,45,572184,679,2,6,281776,1,3618,0.18,603413,1459090,76,16871.0,88,12,5,2,0,1165913,False,False,False,True,False,False,False,False,False,True,False
3,39,2579592,660,5,0,656977,0,16114,0.05,1113302,1362180,112,193469.4,66,4,3,5,0,1869491,True,False,False,False,True,False,False,False,False,True,False
4,58,360612,589,18,0,267892,4,36361,0.32,824809,595409,46,6024.0,70,1,4,2,0,825275,True,True,False,True,False,False,False,False,False,False,False


**Separated Features and Target Variable**

In [9]:
x = df.drop(['Credit_Limit'],axis=1)
y = df['Credit_Limit']

print("Features shape:", x.shape)
print("Target shape:", y.shape)

print("\nTarget Distribution: Credit Limit")
print(y)

Features shape: (5000, 29)
Target shape: (5000,)

Target Distribution: Credit Limit
0       1373925
1       1902449
2       1165913
3       1869491
4        825275
         ...   
4995    1156917
4996    1458578
4997    1342610
4998    1868828
4999    1272502
Name: Credit_Limit, Length: 5000, dtype: int64


**Split Data into Training and Testing Sets**

In [10]:
from sklearn.model_selection import train_test_split
xtrain,xtest,ytrain,ytest = train_test_split(x,y,test_size=0.2, random_state=42)

## **ML Model Fitting & Regression Analysis**
- **Linear Regression**
- **Random Forest Regression**
- **XGBoost Regression**

In [11]:
from sklearn.linear_model import LinearRegression
modellinear= LinearRegression()
modellinear.fit(xtrain,ytrain)
yplinear= modellinear.predict(xtest)
modellinear

LinearRegression()

In [12]:
from sklearn.ensemble import RandomForestRegressor
modelR=RandomForestRegressor(random_state=42)
modelR.fit(xtrain,ytrain)
ypR=modelR.predict(xtest)
modelR

RandomForestRegressor(random_state=42)

In [13]:
from xgboost import XGBRegressor
modelxgbr = XGBRegressor(random_state=42) 
modelxgbr.fit(xtrain, ytrain) 
ypxgbr = modelxgbr.predict(xtest)
modelxgbr

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=True, eval_metric=None, feature_types=None,
             feature_weights=None, gamma=None, grow_policy=None,
             importance_type=None, interaction_constraints=None,
             learning_rate=None, max_bin=None, max_cat_threshold=None,
             max_cat_to_onehot=None, max_delta_step=None, max_depth=None,
             max_leaves=None, min_child_weight=None, missing=nan,
             monotone_constraints=None, multi_strategy=None, n_estimators=None,
             n_jobs=None, num_parallel_tree=None, ...)

# **Metrics Evaluation & 5-Fold Cross-Validation**

### **Linear Regression**

In [14]:
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
Linear_R2score = r2_score(ytest , yplinear)
lrMAE = mean_absolute_error(ytest, yplinear)
lrMSE = mean_squared_error(ytest, yplinear)
lrRMSE = np.sqrt(mean_squared_error(ytest, yplinear))

print("Linear_R2score :", Linear_R2score*100)
print("Mean Absolute Error:", lrMAE)
print("Mean Squared Error:", lrMSE)
print("Root_Mean Squared Error:", lrRMSE)

Linear_R2score : 99.99999999990833
Mean Absolute Error: 0.24945746780140327
Mean Squared Error: 0.08344295570156765
Root_Mean Squared Error: 0.28886494370478333


In [15]:
from sklearn.model_selection import cross_val_score
cvlinear = LinearRegression()
linear_cvscore = cross_val_score(cvlinear,x,y,cv=5,scoring='r2')*100
print(linear_cvscore)
print('linear_cvscore',linear_cvscore.mean())
print("Standard Deviation:",linear_cvscore.std())

[100. 100. 100. 100. 100.]
linear_cvscore 99.9999999999089
Standard Deviation: 4.083345124406858e-12


### **Random Forest Regression**

In [16]:
Random_R2score = r2_score(ytest , ypR)
rfMAE = mean_absolute_error(ytest, ypR)
rfMSE = mean_squared_error(ytest, ypR)
rfRMSE = np.sqrt(mean_squared_error(ytest, ypR))

print("Random_R2score:", Random_R2score*100)
print("Mean Absolute Error:", rfMAE)
print("Mean Squared Error:", rfMSE)
print("Root_Mean Squared Error:", rfRMSE)

Random_R2score: 97.30719046227749
Mean Absolute Error: 39138.72418
Mean Squared Error: 2451005005.982987
Root_Mean Squared Error: 49507.62573566811


In [17]:
cvRandom = RandomForestRegressor()
Random_cvscore = cross_val_score(cvRandom,x,y,cv=5,scoring='r2')*100
print(Random_cvscore)
print('Random_cvscore',Random_cvscore.mean())

[97.4936737  97.37107499 97.44558848 97.28568008 96.93024141]
Random_cvscore 97.30525173145887


### **XGBoost Regression**

In [18]:
XGB_R2score = r2_score(ytest , ypxgbr)
xgMAE = mean_absolute_error(ytest, ypxgbr)
xgMSE = mean_squared_error(ytest, ypxgbr)
xgRMSE = np.sqrt(mean_squared_error(ytest,ypxgbr))

print("XGBoost_R2score:", XGB_R2score*100)
print("Mean Absolute Error:", xgMAE)
print("Mean Squared Error:", xgMSE)
print("Root_Mean Squared Error:", xgRMSE)

XGBoost_R2score: 98.63700866699219
Mean Absolute Error: 27832.048828125
Mean Squared Error: 1240597248.0
Root_Mean Squared Error: 35222.1130541596


In [19]:
cvXGB= XGBRegressor()
XGB_cvscore = cross_val_score(cvXGB,x,y,cv=5,scoring='r2')*100
print(XGB_cvscore)
print('XGBoost_cvscore',XGB_cvscore.mean())

[98.8853395  98.74556065 98.69360328 98.62086773 98.44020605]
XGBoost_cvscore 98.67711544036865


## **Hyperparameter Tuning using GridSearchCV**
- **Random Forest & XGBoost Regression**

In [20]:
from sklearn.model_selection import GridSearchCV
param_grid_rf = {
    'n_estimators': [100, 200],
    'max_depth': [None, 5, 10],
    'min_samples_split': [2, 5]
}

grid_rf = GridSearchCV(
    estimator=modelR,
    param_grid=param_grid_rf,
    cv=5,
    scoring='r2',
    n_jobs=-1)

grid_rf.fit(xtrain,ytrain)
grid_rf.best_estimator_

print(" Random Forest Regression \n")
print("Best Parameters:")
print(grid_rf.best_params_)
print("Best R2 Score:", grid_rf.best_score_*100)

 Random Forest Regression 

Best Parameters:
{'max_depth': None, 'min_samples_split': 2, 'n_estimators': 200}
Best R2 Score: 97.0434863576398


In [21]:
tunned_params_xgb = {'n_estimators': [100, 200],
                         'learning_rate': [0.05, 0.1],
                         'max_depth': [3, 5],
                         'subsample': [0.8, 1.0],
                         'colsample_bytree': [0.8, 1.0]}

params_xgb = GridSearchCV(
    estimator = modelxgbr,param_grid = tunned_params_xgb,scoring='r2',cv=5,n_jobs=-1
)

params_xgb.fit(xtrain, ytrain)
params_xgb.best_estimator_

print("XGBoost Regression \n")
print("Best Parameters:")
print(params_xgb.best_params_)
print("Best R2 Score:", params_xgb.best_score_*100)

XGBoost Regression 

Best Parameters:
{'colsample_bytree': 1.0, 'learning_rate': 0.1, 'max_depth': 3, 'n_estimators': 200, 'subsample': 0.8}
Best R2 Score: 99.53508257865906


### **Model Performance Comparison** 
- **Best Model Selection & Hyperparameter**

In [24]:
results = pd.DataFrame({
    "Model": ['Linear Regression','Random Forest Regressor','XGBoost Regressor'],
    "R2 Score": [Linear_R2score*100,Random_R2score*100,XGB_R2score*100],
    "Cross_Validation_Score": [linear_cvscore.mean(),Random_cvscore.mean(),XGB_cvscore.mean()],
    "MAE": [lrMAE,rfMAE,xgMAE],
    "MSE": [lrMSE,rfMSE,xgMSE ],
    "RMSE": [lrRMSE,rfRMSE,xgRMSE],
})

results = results.sort_values(by=["R2 Score", "Cross_Validation_Score","MAE", "MSE", "RMSE"], ascending=[False, False, True, True, True]).reset_index(drop=True)
results

,Model,R2 Score,Cross_Validation_Score,MAE,MSE,RMSE
0,Linear Regression,100.000000,100.000000,0.249457,8.344296e-02,0.288865
1,XGBoost Regressor,98.637009,98.677115,27832.048828,1.240597e+09,35222.113054
2,Random Forest Regressor,97.307190,97.305252,39138.724180,2.451005e+09,49507.625736


In [25]:
results2 = results.sort_values(
    by='R2 Score',
    ascending=False
).reset_index(drop=True)

results2["Best Parameters"] = results2["Model"].map({
    "Linear Regression": "Default Parameters",
    "Random Forest Regressor": str(grid_rf.best_params_),
    "XGBoost Regressor": str(params_xgb.best_params_)
})

best_model_name = results2.iloc[0]["Model"]

print("Best Model:", best_model_name)

results2

Best Model: Linear Regression


,Model,R2 Score,Cross_Validation_Score,MAE,MSE,RMSE,Best Parameters
0,Linear Regression,100.000000,100.000000,0.249457,8.344296e-02,0.288865,Default Parameters
1,XGBoost Regressor,98.637009,98.677115,27832.048828,1.240597e+09,35222.113054,"{'colsample_bytree': 1.0, 'learning_rate': 0.1..."
2,Random Forest Regressor,97.307190,97.305252,39138.724180,2.451005e+09,49507.625736,"{'max_depth': None, 'min_samples_split': 2, 'n..."


### Model Selection & Performance Insights

- **Linear Regression achieved near-perfect performance (R² ≈ 99.99–100%) with consistent cross-validation results Annual Income (~0.80 correlation) is the strongest predictor of Credit Limit in the dataset**
- **XGBoost (98.64%) and Random Forest (97.31%) also delivered strong predictive performance**
- **However, their higher MAE, MSE, and RMSE indicate greater prediction errors than Linear Regression, 
Linear Regression is selected as the final model due to its accuracy, stability, simplicity, and efficiency**